### PR TITLE
implement a better version of page-scrolling

### PR DESCRIPTION
**Pull request by Guest0x0**
See #2.

This commit implements a better version of page-scrolling, within the original config framework.

It has been tested on my setup, but further test for shell script compatibility would be great.

The README is also updated to notify the difference in behavior.

### CODE REVIEW

**Comment by caksoylar** [comment]
Thanks! I'll leave some suggestions, mostly for nits and to make the linter happy. I also have an idea to preserve the native cursor positioning if the selection needs to change; I can send the suggestion for that after these.

File: smooth-scroll.kak [hunk 1]
```diff
@@ -232,3 +248,21 @@ define-command smooth-scroll-move -params 1 -hidden %{
         printf 'set-option window scroll_running %s\n' "$!"
     }
 }
+
+define-command -hidden -params 1 smooth-scroll-by-page -docstring %{
+    scroll smoothly by (1 / %arg{1}) pages, positive for down, negative for up.
+} %{
+    evaluate-commands %sh{
+        distance=$(expr $kak_window_height / $1)
```
Comment (caksoylar) [inline smooth-scroll.kak]: Was there a specific reason you used `expr`, can we avoid calling it? We can also add support for `count` movements like this:
```suggestion
        if [ "$kak_count" = 0 ]; then
            kak_count=1
        fi
        distance=$(( kak_count * kak_window_height / $1 ))
```

File: smooth-scroll.kak [hunk 2]
```diff
@@ -70,7 +70,23 @@ define-command smooth-scroll-enable -docstring "enable smooth scrolling for wind
                 # in case both sides of the mapping were given with lhs=rhs, split
                 lhs=${key%%=*}
                 rhs=${key#*=}
-                printf 'smooth-scroll-map-key %s %s %s\n' "$mode" "$lhs" "$rhs"
+                case "$mode#$rhs" in
+                    'normal#<c-f>' )
+                        printf 'map window %s %s ":smooth-scroll-by-page 1<ret>"\n' "$mode" "$lhs"
```
Comment (caksoylar) [inline smooth-scroll.kak]:
```suggestion
                        printf 'map window %s %s ": smooth-scroll-by-page 1<ret>"\n' "$mode" "$lhs"
```

File: smooth-scroll.kak [hunk 3]
```diff
@@ -70,7 +70,23 @@ define-command smooth-scroll-enable -docstring "enable smooth scrolling for wind
...
+                        ;;
+                    'normal#<c-b>' )
+                        printf 'map window %s %s ":smooth-scroll-by-page -1<ret>"\n' "$mode" "$lhs"
```
Comment (caksoylar) [inline smooth-scroll.kak]:
```suggestion
                        printf 'map window %s %s ": smooth-scroll-by-page -1<ret>"\n' "$mode" "$lhs"
```

File: smooth-scroll.kak [hunk 4]
```diff
@@ -70,7 +70,23 @@ define-command smooth-scroll-enable -docstring "enable smooth scrolling for wind
...
+                        ;;
+                    'normal#<c-d>' )
+                        printf 'map window %s %s ":smooth-scroll-by-page 2<ret>"\n' "$mode" "$lhs"
```
Comment (caksoylar) [inline smooth-scroll.kak]:
```suggestion
                        printf 'map window %s %s ": smooth-scroll-by-page 2<ret>"\n' "$mode" "$lhs"
```

File: smooth-scroll.kak [hunk 5]
```diff
@@ -70,7 +70,23 @@ define-command smooth-scroll-enable -docstring "enable smooth scrolling for wind
...
+                        ;;
+                    'normal#<c-u>' )
+                        printf 'map window %s %s ":smooth-scroll-by-page -2<ret>"\n' "$mode" "$lhs"
```
Comment (caksoylar) [inline smooth-scroll.kak]:
```suggestion
                        printf 'map window %s %s ": smooth-scroll-by-page -2<ret>"\n' "$mode" "$lhs"
```

File: smooth-scroll.kak [hunk 6]
```diff
@@ -232,3 +248,21 @@ define-command smooth-scroll-move -params 1 -hidden %{
         printf 'set-option window scroll_running %s\n' "$!"
     }
 }
+
+define-command -hidden -params 1 smooth-scroll-by-page -docstring %{
+    scroll smoothly by (1 / %arg{1}) pages, positive for down, negative for up.
+} %{
+    evaluate-commands %sh{
+        distance=$(expr $kak_window_height / $1)
+        if [ $kak_cursor_line -ge $(( ${kak_window_range%% *} + $distance )) ] \
```
Comment (caksoylar) [inline smooth-scroll.kak]:
```suggestion
        if [ "$kak_cursor_line" -ge $(( ${kak_window_range%% *} + distance )) ] \
```

File: smooth-scroll.kak [hunk 7]
```diff
@@ -232,3 +248,21 @@ define-command smooth-scroll-move -params 1 -hidden %{
...
+        && [ $kak_cursor_line -le $(( ${kak_window_range%% *} + $distance + $kak_window_height )) ];
```
Comment (caksoylar) [inline smooth-scroll.kak]:
```suggestion
        && [ "$kak_cursor_line" -le $(( ${kak_window_range%% *} + distance + kak_window_height )) ];
```

File: smooth-scroll.kak [hunk 8]
```diff
@@ -232,3 +248,21 @@ define-command smooth-scroll-move -params 1 -hidden %{
...
+        then
+            # The cursor doesn't need to move, save the selection
+            printf 'set-option window scroll_selections %s\n' "$kak_selections_desc"
+        else
+            # The cursor has to move, don't save it
+            printf 'set-option window scroll_selections\n'
+        fi
+        printf "smooth-scroll-move %s\n" "$distance"
```
Comment (caksoylar) [inline smooth-scroll.kak]:
```suggestion
        printf 'smooth-scroll-move %s\n' "$distance"
```

File: smooth-scroll.kak (hunk 1 above)
Comment (Guest0x0) [reply smooth-scroll.kak]: I remembered `expr` is more expressive than `$((...))`, but I forgot why I chose it in this particular case. After some testing the `expr` call is indeed unnecessary.

The count support is great, thanks!

**Comment by caksoylar** [comment]
I modified the logic a bit so that we can use the regular mapping behavior if the cursor has to move. In my testing this works as expected, but please test and let me know. 

I had to modify `smooth-scroll-by-page` to pass the key as an extra param so we can call `smooth-scroll-do-key` with that, but maybe the interface can be improved.

```diff
diff --git a/smooth-scroll.kak b/smooth-scroll.kak
index 7c20139..b25bc7d 100644
--- a/smooth-scroll.kak
+++ b/smooth-scroll.kak
@@ -72,16 +72,16 @@ define-command smooth-scroll-enable -docstring "enable smooth scrolling for wind
                 rhs=${key#*=}
                 case "$mode#$rhs" in
                     'normal#<c-f>' )
-                        printf 'map window %s %s ": smooth-scroll-by-page 1<ret>"\n' "$mode" "$lhs"
+                        printf 'map window %s %s ": smooth-scroll-by-page 1 <lt>c-f<gt><ret>"\n' "$mode" "$lhs"
                         ;;
                     'normal#<c-b>' )
-                        printf 'map window %s %s ": smooth-scroll-by-page -1<ret>"\n' "$mode" "$lhs"
+                        printf 'map window %s %s ": smooth-scroll-by-page -1 <lt>c-b<gt><ret>"\n' "$mode" "$lhs"
                         ;;
                     'normal#<c-d>' )
-                        printf 'map window %s %s ": smooth-scroll-by-page 2<ret>"\n' "$mode" "$lhs"
+                        printf 'map window %s %s ": smooth-scroll-by-page 2 <lt>c-d<gt><ret>"\n' "$mode" "$lhs"
                         ;;
                     'normal#<c-u>' )
-                        printf 'map window %s %s ": smooth-scroll-by-page -2<ret>"\n' "$mode" "$lhs"
+                        printf 'map window %s %s ": smooth-scroll-by-page -2 <lt>c-u<gt><ret>"\n' "$mode" "$lhs"
                         ;;
                     * )
                         printf 'smooth-scroll-map-key %s %s %s\n' "$mode" "$lhs" "$rhs"
@@ -249,8 +249,10 @@ define-command smooth-scroll-move -params 1 -hidden %{
     }
 }
 
-define-command -hidden -params 1 smooth-scroll-by-page -docstring %{
-    scroll smoothly by (1 / %arg{1}) pages, positive for down, negative for up.
+define-command -hidden -params 2 smooth-scroll-by-page -docstring %{
+    scroll smoothly by (1 / %arg{1}) pages, positive for down, negative for up. if
+    the cursor doesn't have to move, scroll manually. otherwise, emulate the key
+    press given by %arg{2}.
 } %{
     evaluate-commands %sh{
         if [ "$kak_count" = 0 ]; then
@@ -260,12 +262,12 @@ define-command -hidden -params 1 smooth-scroll-by-page -docstring %{
         if [ "$kak_cursor_line" -ge $(( ${kak_window_range%% *} + distance )) ] \
         && [ "$kak_cursor_line" -le $(( ${kak_window_range%% *} + distance + kak_window_height )) ];
         then
-            # The cursor doesn't need to move, save the selection
+            # The cursor doesn't need to move, save the selection and move manually
             printf 'set-option window scroll_selections %s\n' "$kak_selections_desc"
+            printf "smooth-scroll-move %s\n" "$distance"
         else
-            # The cursor has to move, don't save it
-            printf 'set-option window scroll_selections\n'
+            # The cursor has to move, so emulate the key press
+            printf 'smooth-scroll-do-key "%s" ""\n' "$2"
         fi
-        printf 'smooth-scroll-move %s\n' "$distance"
     }
 }
```

**Comment by Guest0x0** [comment]
Oh! When I thought of emulate the vanilla behavior, I thought we must calculate the correct cursor position manually. But yes, you're right, we can just re-use `smooth-scroll-do-key` for that!

It works for me, too. But I'll suggest that you abandon this PR and commit it yourself, since the README change has become unnecessary. I'll leave a link to the commit here and close the PR afterward.

**Comment by caksoylar** [comment]
Since you opened this PR for the v2 branch anyway, I'll merge and then I can revert the README change + add my change on v2. After that I can squash merge it onto the master branch. Thanks again for your contribution!

**Pull request merged**